# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: xpcom/base/CycleCollectedJSRuntime.cpp
# Commit: 687d6e55be66
# Full Hash: 687d6e55be668914ce8f59769d06bedbf1736406
# Author: Matthew Gaudet <mgaudet@mozilla.com>
# Date: 2025-06-03 03:42:51
# Description:
#   Bug 1969445 - Don't report OOM telemetry if the associated global isn't done setting up r=mccr8,jandem
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D251924
# ==============================================================================

diff -r 43ee34b22c20 -r 687d6e55be66 xpcom/base/CycleCollectedJSRuntime.cpp
--- a/xpcom/base/CycleCollectedJSRuntime.cpp	Mon Jun 02 19:24:07 2025 +0000
+++ b/xpcom/base/CycleCollectedJSRuntime.cpp	Mon Jun 02 19:24:12 2025 +0000
@@ -2013,27 +2013,41 @@
   //
   // We may not always collect telemetry, and that's got to be OK :)
   CycleCollectedJSContext* ccjsContext = GetContext();
-  JSContext* jsContext = ccjsContext ? ccjsContext->Context() : nullptr;
-  JSObject* global = jsContext ? JS::CurrentGlobalOrNull(jsContext) : nullptr;
-  if (global) {
-    if (aNewState == OOMState::Recovered) {
-      switch (size) {
-        case Size::Large:
-          SetUseCounter(global, eUseCounter_custom_JS_large_oom_recovered);
-          break;
-        case Size::Small:
-          SetUseCounter(global, eUseCounter_custom_JS_small_oom_recovered);
-          break;
-      }
-    } else {
-      switch (size) {
-        case Size::Large:
-          SetUseCounter(global, eUseCounter_custom_JS_large_oom_reported);
-          break;
-        case Size::Small:
-          SetUseCounter(global, eUseCounter_custom_JS_small_oom_reported);
-          break;
-      }
+  if (!ccjsContext) {
+    return;
+  }
+  JSContext* jsContext = ccjsContext->Context();
+  if (!jsContext) {
+    return;
+  }
+  JS::Realm* realm = JS::GetCurrentRealmOrNull(jsContext);
+
+  // Don't try to report telemetry if the realm is not initialized.
+  if (!realm || !JS::HasRealmInitializedGlobal(realm)) {
+    return;
+  }
+  JSObject* global = JS::GetRealmGlobalOrNull(realm);
+  if (!global) {
+    return;
+  }
+
+  if (aNewState == OOMState::Recovered) {
+    switch (size) {
+      case Size::Large:
+        SetUseCounter(global, eUseCounter_custom_JS_large_oom_recovered);
+        break;
+      case Size::Small:
+        SetUseCounter(global, eUseCounter_custom_JS_small_oom_recovered);
+        break;
+    }
+  } else {
+    switch (size) {
+      case Size::Large:
+        SetUseCounter(global, eUseCounter_custom_JS_large_oom_reported);
+        break;
+      case Size::Small:
+        SetUseCounter(global, eUseCounter_custom_JS_small_oom_reported);
+        break;
     }
   }
 }
